src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java

```diff
@@ -87,8 +88,8 @@ public DefaultClientCredentialsOAuthDancer(String clientId, String clientSecret,
   @Override
   public void start() throws MuleException {
     super.start();
-    // We use a reentrant instead of one from the lock factory because the local state of this object cannot be shared in the
-    // cluster.
+    // TODO MULE-17013 We use a reentrant instead of one from the lock factory because the local state of this object cannot be
```

Comment:
this was already done. We used to have a monitored object store in which we put an entry indicating that ta refresh token was in place. That entry had a relatively short TTL. That ObjectSTore was checked before dooing the refresh and if it was not empty, the refresh was not attempted and the current token was returned instead

Reply:
that was ok for a blocking implementation, where you could just wait for the token to be put into the OS. For this non-blocking implementation, we need a notification to complete the returned `CompletableFuture`.

Reply:
what about a VM queue? Also, keep cloudhub limitations when defining this. We already have issues with the distributed locking. Need to come out with a solutiono that also works there

Reply:
a vm queue won't work with a cluster bigger than 2 nodes. Agree we need to further discuss the situation with CH.